src/server/service/tracing.rs

```diff
@@ -0,0 +1,338 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::sync::Arc;
+// use std::lazy::Lazy;
+use std::time::Duration;
+
+use arc_swap::ArcSwap;
+use futures::channel::mpsc::{channel, Receiver, Sender};
+use futures::prelude::*;
+use kvproto::tracepb;
+use minitrace::prelude::*;
+use online_config::{ConfigChange, OnlineConfig};
+use serde_json::json;
+use tikv_util::config::ReadableDuration;
+use tikv_util::time::{duration_to_sec, Instant};
+
+use crate::server::metrics::{GrpcTypeKind, GRPC_MSG_HISTOGRAM_STATIC};
+
+type CollectTask = (Collector, Vec<tracepb::RemoteParentSpan>, Duration);
+
+// static SPAN_ID_RANDOM_BITS: Lazy<u32> = Lazy::new(|| rand::thread_rng().gen());
+const CHANNEL_SIZE: usize = 1024;
+
+pub fn init_tracing(
+    config: &TracingConfig,
+) -> (TracingService, TracerFactory, TracingConfigManager) {
+    let config = Arc::new(ArcSwap::from_pointee(config.clone()));
+    let (tx, rx) = channel(CHANNEL_SIZE);
+
+    (
+        TracingService { rx },
+        TracerFactory {
+            config: config.clone(),
+            tx,
+        },
+        TracingConfigManager { config },
+    )
+}
+
+pub struct TracingService {
+    rx: Receiver<CollectTask>,
+}
+
+impl TracingService {
+    // TODO (andylokandy): implement a real pubsub service
+    pub fn register_service(self) {
+        std::thread::spawn(move || {
+            let print_to_slow_log = self.rx.for_each_concurrent(
```

Comment:
Is this log output planned to be removed once ngm endpoint is implemented?

Reply:
yes